mpi-proxy-split/mpi-wrappers/mpi_cart_wrappers.cpp

```diff
@@ -222,6 +161,93 @@ USER_DEFINED_WRAPPER(int, Dims_create, (int)nnodes, (int)ndims, (int *)dims)
+// checkpoint (DMTCP_EVENT_PRECHECKPOINT event in mpi_plugin.cpp).
+CartesianProperties g_cartesian_properties = { .comm_old_size = -1,
+                                               .comm_cart_size = -1,
+                                               .comm_old_rank = -1,
+                                               .comm_cart_rank = -1 };
+
+USER_DEFINED_WRAPPER(int, Cart_create, (MPI_Comm)old_comm, (int)ndims,
+                     (const int *)dims, (const int *)periods, (int)reorder,
+                     (MPI_Comm *)comm_cart)
+{
+  //JASSERT(g_cartesian_properties.comm_old_size == -1)
+  //  .Text("MPI_Cart_create() called more than once. Current implementation "
+  //        "only supports one cartesian communicator.");
+
+  std::function<int()> realBarrierCb = [=]() {
+    int retval;
+    DMTCP_PLUGIN_DISABLE_CKPT();
+    MPI_Comm realComm = VIRTUAL_TO_REAL_COMM(old_comm);
+    JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+    retval = NEXT_FUNC(Cart_create)(realComm, ndims, dims, periods, reorder,
+                                    comm_cart);
+    RETURN_TO_UPPER_HALF();
+    g_cartesian_properties.ndims = ndims;
+    g_cartesian_properties.reorder = reorder;
+    for (int i = 0; i < ndims; i++) {
+      g_cartesian_properties.dimensions[i] = dims[i];
+      g_cartesian_properties.periods[i] = periods[i];
+    }
+    MPI_Comm_size(old_comm, &g_cartesian_properties.comm_old_size);
+    MPI_Comm_size(*comm_cart, &g_cartesian_properties.comm_cart_size);
+    MPI_Comm_rank(old_comm, &g_cartesian_properties.comm_old_rank);
+    MPI_Comm_rank(*comm_cart, &g_cartesian_properties.comm_cart_rank);
+    MPI_Cart_coords(*comm_cart, g_cartesian_properties.comm_cart_rank,
+                    g_cartesian_properties.ndims,
+                    g_cartesian_properties.coordinates);
+
+    if (retval == MPI_SUCCESS && MPI_LOGGING()) {
+      MPI_Comm virtComm = ADD_NEW_COMM(*comm_cart);
+      VirtualGlobalCommId::instance().createGlobalId(virtComm);
+      *comm_cart = virtComm;
+      active_comms.insert(virtComm);
+
+      FncArg ds = CREATE_LOG_BUF(dims, ndims * sizeof(int));
+      FncArg ps = CREATE_LOG_BUF(periods, ndims * sizeof(int));
+      LOG_CALL(restoreCarts, Cart_create, old_comm, ndims, ds, ps, reorder,
+               virtComm);
+    }
+    DMTCP_PLUGIN_ENABLE_CKPT();
+    return retval;
+  };
+  return twoPhaseCommit(old_comm, realBarrierCb);
+}
+#else
+
+USER_DEFINED_WRAPPER(int, Cart_create, (MPI_Comm) old_comm, (int) ndims,
+                     (const int*) dims, (const int*) periods, (int) reorder,
+                     (MPI_Comm *) comm_cart)
+{
+  int retval;
+  // We are forcing reordering to be false because in order to ensure that on
+  // restart, each rank has the same cartesian coordinates for each of
+  // communicator that it is part of. This may cause some runtime overhead
+  // on restart on applications that call Cart_create with reorder=true
```

Comment:
Please change the comment like following:
```
The MPI library assigns the cartesian coordinates naively (in the increasing rank order) with no reordering as opposed to optimal reordering based on the physical topology.
For example, if there are 6 ranks (0-5) exist, then without reordering, the two-dimensional coordinates will be following:
(0, 0) -> rank 0
(0, 1) -> rank 1
.
.
(2, 2) -> rank 5
While no reordering ensures the same rank to coordinates mapping on the restart,  it can incur some overhead as MPI can no longer optimize the ranks' ordering. For now, we focus on the correctness across checkpoint-restart. Therefore, we enforce the reorder variable to be false.

FIXME: Handle the case when reorder variable is true.
```
and add a warning message whenever reorder is true:
```
JWARNING (reorder == false) .Text ("We are enforcing reorder to false as the current implementation does not support reordered ranks.");
```